apps/typegpu-docs/src/content/examples/algorithms/mnist-inference/index.ts

```diff
@@ -0,0 +1,390 @@
+      }),
+    ),
+  );
+
+  const weightsBindGroups = buffers.map((layer) =>
+    root.unwrap(
+      weightsBiases.populate({
+        weights: layer.buffers.weights,
+        biases: layer.buffers.biases,
+      }),
+    ),
+  );
+
+  async function inference(data: number[]): Promise<number[]> {
+    // verify the length of the data matches the input layer
+    if (data.length !== layers[0][0].shape[0]) {
+      throw new Error(
+        `Data length ${data.length} does not match input shape ${layers[0][0].shape[0]}`,
+      );
+    }
+    input.write(data);
+
+    // Run the network
+    const encoder = device.createCommandEncoder();
+    for (let i = 0; i < buffers.length; i++) {
+      const pass = encoder.beginComputePass();
+      pass.setPipeline(pipeline);
+      pass.setBindGroup(0, ioBindGroups[i]);
+      pass.setBindGroup(1, weightsBindGroups[i]);
+      pass.dispatchWorkgroups(buffers[i].biases.shape[0]);
+      pass.end();
+    }
+    device.queue.submit([encoder.finish()]);
+    await device.queue.onSubmittedWorkDone();
+
+    // Read the output
+    return await output.read();
+  }
+
+  return {
+    layers: buffers,
+    input,
+    output,
+    inference,
+  };
+}
+
+// Data fetching and network creation
+
+const layer0Biases = await fetch('/TypeGPU/mnistWeights/layer0.bias.npy').then(
+  (res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+);
+const layer0Weights = await fetch(
+  '/TypeGPU/mnistWeights/layer0.weight.npy',
+).then((res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)));
+
+const layer1Biases = await fetch('/TypeGPU/mnistWeights/layer1.bias.npy').then(
+  (res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+);
+const layer1Weights = await fetch(
+  '/TypeGPU/mnistWeights/layer1.weight.npy',
+).then((res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)));
+
+const layer2Biases = await fetch('/TypeGPU/mnistWeights/layer2.bias.npy').then(
+  (res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+);
+const layer2Weights = await fetch(
+  '/TypeGPU/mnistWeights/layer2.weight.npy',
+).then((res) => res.arrayBuffer().then((buffer) => getLayerData(buffer)));
+
+const network = createNetwork([
+  [layer0Weights, layer0Biases],
+  [layer1Weights, layer1Biases],
+  [layer2Weights, layer2Biases],
+]);
+
+// Canvas drawing
+
+const context = canvas.getContext('2d') as CanvasRenderingContext2D;
+
+const bars = Array.from(
+  { length: 10 },
+  (_, i) => document.getElementById(`bar-fill-${i}`) as HTMLElement,
+);
+
+const resetAll = () => {
+  data.fill(0);
+  resetCanvas();
+};
+
+const resetCanvas = () => {
+  context.clearRect(0, 0, canvas.width, canvas.height);
+  // draw grid
+  context.strokeStyle = '#ccc';
+  const scale = canvas.width / 28;
+  for (let i = 0; i < 28; i++) {
+    for (let j = 0; j < 28; j++) {
+      context.strokeRect(j * scale, i * scale, scale, scale);
+    }
+  }
+};
+resetCanvas();
+
+const draw = () => {
+  const size = 28;
```

Comment:
extract as global constant and use in resetCanvas and mouse move listener as well